lang/csharp/src/apache/test/Reflect/TestLogMessage.cs

```diff
@@ -0,0 +1,123 @@
+        Error
+    }
+
+    public class LogMessage
+    {
+        private Dictionary<string, string> _tags = new Dictionary<string, string>();
+
+        public string IP { get; set; }
+
+        public string Message { get; set; }
+
+        [Avro(typeof(DateTimeOffsetToLongConverter))]
+        public DateTimeOffset TimeStamp { get; set; }
+
+        public Dictionary<string, string> Tags { get => _tags; set => _tags = value; }
+
+        public MessageTypes Severity { get; set; }
+    }
+
+    public class LogMessage2
+    {
+        public string IP { get; set; }
+
+        public string Message { get; set; }
+
+        public DateTimeOffset TimeStamp { get; set; }
+
+        private Dictionary<string, string> _tags = new Dictionary<string, string>();
+
+        public Dictionary<string, string> Tags { get => _tags; set => _tags = value; }
+
+        public MessageTypes Severity { get; set; }
+    }
+
+    [TestFixture]
+    public class TestLogMessage
+    {
+        private string _logMessageSchemaV1 = @"
+        {
+            ""namespace"": ""MessageTypes"",
+            ""type"": ""record"",
+            ""doc"": ""A simple log message type as used by this blog post."",
+            ""name"": ""LogMessage"",
+            ""fields"": [
+                { ""name"": ""IP"", ""type"": ""string"" },
+                { ""name"": ""Message"", ""type"": ""string"" },
+                { ""name"": ""TimeStamp"", ""type"": ""long"" },
+                { ""name"": ""Tags"",""type"":
+                    { ""type"": ""map"",
+                        ""values"": ""string""},
+                        ""default"": {}},
+                { ""name"": ""Severity"",
+                ""type"": { ""namespace"": ""MessageTypes"",
+                    ""type"": ""enum"",
+                    ""doc"": ""Enumerates the set of allowable log levels."",
+                    ""name"": ""LogLevel"",
+                    ""symbols"": [""None"", ""Verbose"", ""Info"", ""Warning"", ""Error""]}}
+            ]
+        }";
+
+        [TestCase]
+        public void Serialize()
+        {
+            var schema = global::Avro.Schema.Parse(_logMessageSchemaV1);
+            var avroWriter = new ReflectWriter<LogMessage>(schema);
+            var avroReader = new ReflectReader<LogMessage>(schema, schema);
+
+            byte[] serialized;
+
+            var logMessage = new LogMessage()
+            {
+                IP = "10.20.30.40",
+                Message = "Log entry",
+                Severity = MessageTypes.Error
+            };
+
+            using (var stream = new MemoryStream(256))
+            {
+                avroWriter.Write(logMessage, new BinaryEncoder(stream));
+                serialized = stream.ToArray();
+            }
+
+            LogMessage deserialized = null;
+            using (var stream = new MemoryStream(serialized))
+            {
+                deserialized = avroReader.Read(default(LogMessage), new BinaryDecoder(stream));
```

Comment:
We should probably assert that all the properties of `deserialized` match up with those of `logMessage`.

Reply:
Done